File: server/js/websocket_impl.js
const IsoWebSocket = require('isomorphic-ws');

class WebSocketServer {
  start(port, onConnection) {
    this.binding = new IsoWebSocket.Server({ port: port });
    this.binding.on('connection', function (socket, request) {
      onConnection(
        new WebSocket(socket),
        {
          'address': request.connection.remoteAddress,
          'port': request.connection.remotePort,
        }
      );
    });
  }

  close() {
    this.binding.close();
  }
}

class WebSocket {
  constructor(socket) {
    this.socket = socket;
  }

  send(data) {
    this.socket.send(data);
  }

  listen(onMessage) {
    this.socket.on('message', function message(data) {
      onMessage(data);
    });
  }

  close() {
    this.socket.close();
  }
}

module.exports = {
  WebSocketServer,
  WebSocket
};

File: server/dart/lib/comms/websocket/websocket_impl.dart
@JS()
library websocket_js_impl;

import 'package:multiplayer_asteroids_common/common.dart';
import 'package:js/js.dart';

@JS()
class WebSocketServer implements CommsServer {
  external factory WebSocketServer();
  external void start(
    int port,
    void onConnection(WebSocket socket, ClientInfo client),
  );
  external void close();
}

@JS()
class WebSocket implements Socket {
  external factory WebSocket();
  external void send(data);
  external void listen(void onMessage(data));
  external void close();
}

@JS()
@anonymous
class WebSocketClientInfo implements ClientInfo {
  external String get address;
  external int get port;
  external factory WebSocketClientInfo({
    String address,
    int port,
  });
}
